analyzers/src/SonarAnalyzer.CSharp/Rules/LoggerFieldsShouldBePrivateStaticReadonly.cs

```diff
@@ -0,0 +1,92 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class LoggerFieldsShouldBePrivateStaticReadonly : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S1312";
+    private const string MessageFormat = "Make the logger '{0}' private static readonly.";
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+    private static readonly KnownAssembly[] LoggingFrameworks =
+        [
+            KnownAssembly.MicrosoftExtensionsLoggingAbstractions,
+            KnownAssembly.NLog,
+            KnownAssembly.Serilog,
+            KnownAssembly.Log4Net,
+            KnownAssembly.CastleCore,
+        ];
+
+    private static readonly ImmutableArray<KnownType> Loggers = ImmutableArray.Create(
+        KnownType.Microsoft_Extensions_Logging_ILogger,
+        KnownType.Microsoft_Extensions_Logging_ILogger_TCategoryName,
+        KnownType.NLog_ILogger,
+        KnownType.NLog_ILoggerBase,
+        KnownType.NLog_Logger,
+        KnownType.Serilog_ILogger,
+        KnownType.log4net_ILog,
+        KnownType.log4net_Core_ILogger,
+        KnownType.Castle_Core_Logging_ILogger);
+
+    private static readonly HashSet<SyntaxKind> InvalidAccessModifiers =
+        [
+            SyntaxKind.ProtectedKeyword,
+            SyntaxKind.InternalKeyword,
+            SyntaxKind.PublicKeyword
+        ];
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterCompilationStartAction(cc =>
+        {
+            if (cc.Compilation.ReferencesAny(LoggingFrameworks))
+            {
+                cc.RegisterNodeAction(c =>
+                {
+                    foreach (var invalid in InvalidFields((FieldDeclarationSyntax)c.Node, c.SemanticModel))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, invalid.GetLocation(), invalid.ValueText));
+                    }
+                },
+                SyntaxKind.FieldDeclaration);
+            }
+        });
+
+    private static IEnumerable<SyntaxToken> InvalidFields(FieldDeclarationSyntax field, SemanticModel model)
+    {
+        var isStatic = field.Modifiers.Any(x => x.IsKind(SyntaxKind.StaticKeyword));
+        var isReadonly = field.Modifiers.Any(x => x.IsKind(SyntaxKind.ReadOnlyKeyword));
+        var isPrivate = field.Modifiers.All(x => !x.IsAnyKind(InvalidAccessModifiers));
+
+        if (isStatic && isReadonly && isPrivate)
+        {
+            yield break;
+        }
```

Comment:
pedantic: moving the conditions inside the if would be more performance due to the short circuiting behavior:
```csharp
        if (field.Modifiers.Any(x => x.IsKind(SyntaxKind.StaticKeyword))
            && field.Modifiers.Any(x => x.IsKind(SyntaxKind.ReadOnlyKeyword))
            && field.Modifiers.All(x => !x.IsAnyKind(InvalidAccessModifiers)))
        {
            yield break;
        }
```

Reply:
Usually this would be true, but all of these are deferred execution LINQ expressions, so when I declare them as local variables, no computation happens until evaluated.

If I am not missing something, I think the added readability has a marginal benefit.

Reply:
There is no deferred execution here :)

Reply:
# /facepalm 🤦‍♂️ 

...you are right, refactoring it.